### ISSUE TITLE
no flow feature extractor was generated after training

### ISSUE THREAD

**Issue by chris201286** [jbohnslav/deepethogram#32]
There are some issues during training:
1. To use 'deg_s' rather than the default 'deg_f', I have to modify the config.yaml. I think this is quite inconvenient.

2. There is no option for me to choose the new trained flow generator in the Model selector under the Flow generator box (the default was still the downloaded flow generator), although the flow generator was successfully generated. I have to modify the project_config file to use this new trained flow generator.

3. After training feature extractor, no flow feature extractor was generated. There was even no 'flow' folder under the 'feature_extractor_train_None' directory, although spatial feature extractor was successfully generated  ('checkpoint.pt' was deleted from the attached due to space limitation). Therefore, my training had to stop there.

4. I am quite confused about the style of 'path'. Taking 'config_file: C:/Users/ChenLab/Documents/deepethogram/rear_scratch_freeze/rear_scratch_freeze_deepethogram\project_config.yaml' as an example, why is there both '/' and '\' in this statement?

[project_config.txt](https://github.com/jbohnslav/deepethogram/files/5751688/project_config.txt)

[201228_193851_feature_extractor_train_None.zip](https://github.com/jbohnslav/deepethogram/files/5751699/201228_193851_feature_extractor_train_None.zip)

**Comment by mmh513** [comment]
Hello,

For point 1, instead of changing each individual parameter in the project configuration file you can run a command line instead. For example to train a flow generator on deg_s you would open 'cmd.exe' window, run 'activate conda deg' then run 'python -m deepethogram.flow_generator.train project.config_File=c:/path/to/your/project_config.yaml preset= deg_s'. You will find then if you go into the project_config file in the folder of your now newly training flow generator, you should see all the presets for deg_s. I find command line is a very convenient way to use this software.  If you want more information on how to run command lines, see the CLI page in docs on GitHub. 

For point 2/3, as you know within each project folder, there is a project configuration file. Before you open the GUI, go into this file and you will see there is a flow generator section. At the bottom of this section there is a weights line: 


![image](https://user-images.githubusercontent.com/76814178/103443709-e25f4380-4c2f-11eb-8d2b-0be26829f435.png)

You have to update this line to the 'checkpoint.pt' file _for your specific_ trained flow generator folder. Save and close out the project configuration file, then open the GUI and you should see this as one of the options in the drop down menu under flow generator. Similarly, if you want to open a specific feature extractor in the GUI, you would edit the 'weights' line under the feature extractor section in the configuration file and save prior to opening the GUI. 

Good luck friend!

-Matt

**Comment by chris201286** [comment]
Hi Matt,
Thanks for the help.
For point 2, I actually did as you wrote. I modified the project_config.yaml file by adding 'flow_generator:' and then the sefl-trained flow generator could be selected. One thing I noticed was that there were no so many parameters regarding the flow generator in my project_config.yaml file:
![image](https://user-images.githubusercontent.com/76747278/103445174-e4bca000-4c25-11eb-8051-0c18039dcda8.png)

For point 3, the issues was not that l can't select the self-trained feature extractor, the main problem is that no 'checkpoint.pt'' file was generated after the feature extractor training. I noticed that the flow feature extractor was not generated as well:
![image](https://user-images.githubusercontent.com/76747278/103445247-95c33a80-4c26-11eb-985d-8be537215061.png).

Thank you for your further advice in advance,
Chris

**Comment by mmh513** [comment]
Hello, 

As you know, when using deepethogram you will have 1) a command prompt open and 2) the graphic user interface open. When you trained the model above, what did _the command prompt_ say at the end of training? It should say something along the lines of 'learning criteria reached!...etc etc'. This window should not close out. Or did you come back to your computer after starting this training (maybe running it overnight) to find that command prompt was closed out?  From the looks of it (disclaimer I have only used the software and am not part of the development of the software at all), it seems like your command prompt and model closed before the training was complete. This could be due to model complexity. This was also happening to me. Ultimately, I found that my computer itself was having a power failure and shutting off/restarting while training models. I decreased the 'batch size' in my configuration file all the way down to 4. Additionally, I downloaded the open source free software called 'Open Hardware Monitor' and found that some of my fans were only set to 50% capacity. After decreasing my batch size and increasing my fan speed, my models have been training to completion with no issues. Hope this helps!

Sincerely, 

Matt

**Comment by chongtianyifa** [comment]
Hi Matt,

Thanks for your response, especially considering you are not part of the development. I indeed saw 'learning criteria reached!...etc etc'' after trainning flow generator (it took me 2 days and nights for the 'deg_s' model). However, I did not see the 'learning criteria reached!' after trainning the feature extractor and this step ran too quickly (done in about 5 hrs). The command prompt and GUI did not close when I checked the other day. I observed that there were a lot of 'weight not found in model dictionary' warnings when loading the retrained feature extractor during the initiation of training (see the 'train' txt file). I am not sure whether it was the issue to cause the failure of training. However, my thought is that there are some bugs for this pacakge as the 'flow' folder should be created immediately as the 'spatial' folder right after clicking the training button. Hope someone can help me fix this problem.

Best,
Lei

**Comment by jbohnslav** [comment]
Hi @mmh513, thanks a lot for your help with this issue. Sorry I've been absent, I've been working on revisions for the paper. 

> To use 'deg_s' rather than the default 'deg_f', I have to modify the config.yaml. I think this is quite inconvenient.

I have no plans to make the GUI easier to use than changing one line of a text file. 

> There is no option for me to choose the new trained flow generator in the Model selector under the Flow generator box (the default was still the downloaded flow generator), although the flow generator was successfully generated. I have to modify the project_config file to use this new trained flow generator.

This is true. When revisions are in, I can work on making this section of the GUI automatically update.

Note to all (@chongtianyifa, @chris201286, @mmh513): if you use the command line, e.g. `python -m deepethogram.feature_extractor.train` instead of using the `train` button on the GUI, all you have to do is use `flow_generator.weights=latest`, and equivalently `feature_extractor.weights=latest`.  Using the flag `latest`, instead of manually specifying it in your config file, the software will automatically find the latest weights with the same architecture (e.g. deg_f). 

However, if you're training the feature extractor for the first time, please use the pretrained weights directory found on the [getting started page](https://github.com/jbohnslav/deepethogram/blob/25b784757eb11c6a7981e45354e509ff9e028bd3/docs/getting_started.md#download-models). 

> After training feature extractor, no flow feature extractor was generated. There was even no 'flow' folder under the 'feature_extractor_train_None' directory, although spatial feature extractor was successfully generated ('checkpoint.pt' was deleted from the attached due to space limitation). Therefore, my training had to stop there.

There are a number of issues indicated in your log files @chris201286. One: because you are manually specifying weight files, it looks like you've tried to reload weights from a different architecture. Please use the `latest` flag at the command line, or make sure on the GUI that the weight directory you specify corresponds to the same `preset, e.g. deg_s`. Furthermore, it looks like training is stopped very quickly-- there is some error in training. This is why there is no flow folder, etc. What is the full traceback in your terminal window?

> I am quite confused about the style of 'path'. Taking 'config_file: C:/Users/ChenLab/Documents/deepethogram/rear_scratch_freeze/rear_scratch_freeze_deepethogram\project_config.yaml' as an example, why is there both '/' and '' in this statement?

This is a bug. Windows pathing (using `os.path`) is kind of a pain. I'll try to find where this was introduced.

**Comment by mmh513** [comment]
Hi Lei, 

I looked at your training file and found an issue. The reason you are getting so many warnings during feature extractor training is because the trained flow generator parameters were set to deg_f as the arch in your config file was 'TinyMotionNet' instead of 'TinyMotionNet3d'. Here is a screen shot from your config file: 
![image](https://user-images.githubusercontent.com/76814178/103539054-5e8e9c80-4e65-11eb-9559-a6c665b42759.png)

The remedy is simple, you will have to go into the project configuration file for this specific project and change these 5 parameters under flow generator: 
![image](https://user-images.githubusercontent.com/76814178/103540848-63088480-4e68-11eb-91e7-2a032e8b0bb4.png)
After you have saved these updates you can open a command prompt/deepethogram GUI, once you open the project you should see under the flow generator there are the weights for the correct 'TinyMotionNet3D' which is needed for Deg_s. Additionally I noticed when looking again at your project_config file that under feature extractor  this is also set for deg_f as the arch is resnet18 instead of resnet_3d_34. Before training the feature extractor you will again have to update the project config file.   Hope this helps!

-Matt

**Comment by chongtianyifa** [comment]
Hi Matt and Jim,

Thank you for your kind help. After reading the discussion above, I decided to retrain from the scratch using the command line. For the flow generator, it looks correct now.
![image](https://user-images.githubusercontent.com/46367566/103604112-07201900-4ec5-11eb-9767-d730b8c5f88d.png)
No more 'weight not found in model dictionary' warnings. I will report the training progress if something goes wrong.

Thanks again,
Lei

**Comment by chris201286** [comment]
Hi Matt and Jim,

By using the command line, I successfully trained the flow generator for deg_s.
![image](https://user-images.githubusercontent.com/76747278/103790517-29658400-4ff6-11eb-9819-06892ce046b0.png)
For the feature extractor training, the spatial feature was also trained.
![image](https://user-images.githubusercontent.com/76747278/103790718-72b5d380-4ff6-11eb-99c4-ca5977bbc8af.png)
However, some error occurred within 2 hrs after the initiation of flow feature training due to CUDA error.
![image](https://user-images.githubusercontent.com/76747278/103791065-d9d38800-4ff6-11eb-8802-09ffed0ac8f2.png)
Is there any way I can restart the training using the trained spatial feature rather than training the feature extractor from the scratch?
![image](https://user-images.githubusercontent.com/76747278/103791392-43ec2d00-4ff7-11eb-82ac-a647b9ea437b.png)
The issue is that I did not find a checkpoint.pt file under the feature_extractor_train folder, although there are such files under the 
flow and spatial subfolders.

Another issue was that the loss was very high for 'val' data while it is low for 'train' data during spatial feature. Does it mean the model was overfit? Meanwhile, the f1 was very small during spatial feature, is that normal as the incoming flow feature training will help decrease the loss and increase the f1 for the 'val' data?
![image](https://user-images.githubusercontent.com/76747278/103791562-73029e80-4ff7-11eb-9824-ae2be9aec0a9.png)

**Comment by chris201286** [comment]
Hi,

I tried the feature extractor training again and it stopped at about 1h after the initiation of flow feature training due to a CUDA error again.
![image](https://user-images.githubusercontent.com/76747278/103913753-c17a7080-50bd-11eb-8b7f-e15733b5920e.png)
Does someone know what the issue might be?

Thanks,
Lei

**Comment by jbohnslav** [comment]
Hmm, these kinds of CUDA errors are notoriously hard to debug. What version of PyTorch / CUDA-toolkit are you using? What GPU?

**Comment by chris201286** [comment]
PyTorch: 1.7.1     py3.7_cuda102_cudnn7_0  (build channel)
cudatoolkit: 10.2.89              h74a9793_1 (build channel)
My GPU is RTX 2080 Ti.

**Comment by jbohnslav** [comment]
All that looks OK. Because it's a non-specific CUDA bug, it's not really something I can debug on my end... is it possible the GPU is overheating? 

PS a lot of these issues are likely to be improved by just using DEG_m instead of DEG_s.

**Comment by chris201286** [comment]
Hi Jim,
I tried the deg_m and got this error during the  flow feature training after about 17 epochs.
![image](https://user-images.githubusercontent.com/76747278/104497247-eb84d480-558e-11eb-9b26-f5317a8db6ce.png)

**Comment by jbohnslav** [comment]
Hi Chris,

Luckily, this one is fixable-- could you uninstall and reinstall from source? This bug has been fixed. I'm waiting to make a big update before I make a new Pip version.

**Comment by chris201286** [comment]
Hi Jim,

That's great. Could you give some guides to uninstall and reinstall from source? I installed deg following this step rather than from source:
![image](https://user-images.githubusercontent.com/76747278/104498177-3226fe80-5590-11eb-8473-6591b07f5457.png)

**Comment by jbohnslav** [comment]
Sure, check this readme https://github.com/jbohnslav/deepethogram/blob/master/docs/installation.md

It is basically just 
* pip uninstall deepethogram
* download latest from Git (instructions above)
* cd to code directory
* python setup.py develop

**Comment by chris201286** [comment]
Hi Jim,

I reinstalled deepethogram using the latest version. Now I successfully trained the feature extractor and inferred the feature for these trained videos. However, I got errors during training the 'sequence model' using these inferred features.
![image](https://user-images.githubusercontent.com/76747278/105118521-711eed80-5a83-11eb-92cc-073daee1d784.png)
Can you help me fix this issue?

Thanks,
Chris

**Comment by chongtianyifa** [comment]
Hi Jim,

I have two questions about the function of the flow generator and sequence model.
1. what is the function of flow generator?
From my understanding, it is used to generate optic flow information. According to the paper said "Also, we trained the flow generator only once and kept it fixed for all experiments.", Is the training of flow generator used to adjust the weights of MotionNet architectures to better fit specific experiment environment? Should we run the flow generator for newly added videos to be predicted if the experimental condition changes?

2. Although the running of train/infer for the sequence model hasn't worked for me as well, I can already see the prediction results. Does this mean that the sequence model is used to optimize the results output by infer in feature extractor? 

If my understanding is correct, the workflow might look better in this way:
![image](https://user-images.githubusercontent.com/46367566/105137236-dc7ab680-5aa7-11eb-81c7-32fbb25664a6.png)
Rather than you previous post:
![image](https://user-images.githubusercontent.com/46367566/105137293-eac8d280-5aa7-11eb-9655-7d670b362ae3.png)

Thanks,
Lei

**Comment by jbohnslav** [comment]
Hi @chris201286 , could you post the full stack trace? For some reason it is trying to reload a previous model, despite the fact that you have not trained a sequence model yet. 

@chongtianyifa: 

> what is the function of flow generator?
From my understanding, it is used to generate optic flow information. According to the paper said "Also, we trained the flow generator only once and kept it fixed for all experiments.", Is the training of flow generator used to adjust the weights of MotionNet architectures to better fit specific experiment environment? Should we run the flow generator for newly added videos to be predicted if the experimental condition changes?

This is correct, we are updating the weights used to calculate optic flow to make it more suitable to your specific videos. If you are satisfied with the optic flow performance (viewed currently as a tiff stack of figures in your model run directory) then you can just keep that fixed. If you change the videos dramatically (moving from grayscale to color, moving to a condition where there is now bedding in the cage instead of a white background) you should re-train it. 

> Although the running of train/infer for the sequence model hasn't worked for me as well, I can already see the prediction results. Does this mean that the sequence model is used to optimize the results output by infer in feature extractor?

If there are problems with the sequence training, I should fix those. If you look at figure 5G, sequence models consistently improve the predictions. However, if your labels are good enough at the intermediate step of the predictions in the feature extractor, of course you can stop there.

**Comment by chris201286** [comment]
Hi Jim,

I don't get 'the full stack trace'. Do you mean the full prompt information?
![image](https://user-images.githubusercontent.com/76747278/105759454-5aeeb280-5f05-11eb-96d2-d2e3b9df8ff2.png)

[train.log](https://github.com/jbohnslav/deepethogram/files/5869349/train.log)
[config.txt](https://github.com/jbohnslav/deepethogram/files/5869350/config.txt)

**Comment by jbohnslav** [comment]
@chris201286 are you using the command line or the GUI? It is very odd that `reload.latest` is True even though you haven't trained any models yet. You can do one of two things: from the command line, add the flag `reload.latest=False`. Or you can download the latest code in the master branch and install it-- with commit e9154c4, trying to do this will only result in a warning, not an error.

**Comment by chris201286** [comment]
Hi Jim,
Thanks for the advice. I used the command line for these trainings. It is now running by set the flag reload.latest=False, although there is warining:
![image](https://user-images.githubusercontent.com/76747278/106044315-8ac2d380-611a-11eb-8fb9-e6b633a65805.png)
I think this warning is normal as no pre-trained sequence model. Do I need to set the reload.latest=true for future further sequence training after this?

**Comment by jbohnslav** [comment]
Honestly, for the sequence models it shouldn't matter much to reload the weights or not. The really important ones to reload are the flow generator and feature extractor-- without proper weights reloading, it could take a day or two of training. For the sequence model, it should be much faster.

**Comment by chongtianyifa** [comment]
Hi Jim,

I have trained the flow generator/feature extractor/sequence model. After using this trained model to predict a new video, I found a significant issue. The prediction is good for videos with the experimental environment similar to that in the trained videos. However, it was quite bad for novel videos recorded in new condition (e.g, mice were attached to an optic fiber in new videos while mice were not in the trained video). The possible explanation is that the motion of optic fiber disrupted the motion detection by the flow generator, thus impairing the subsequent extraction of the flow feature. Will this issue be fixed by adding more videos in which mice are connected to optic fiber and retraining the model?

Thanks,
Chris
